### PR TITLE
[SPARK-32163][SQL] Nested pruning should work even with cosmetic variations

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to deal with cosmetic variations when processing nested column extractors in `NestedColumnAliasing`. Currently if cosmetic variations are in the nested column extractors, the query is not optimized.

### Why are the changes needed?

If the expressions extracting nested fields have cosmetic variations like qualifier difference, currently nested column pruning cannot work well.

For example, two attributes which are semantically the same, are referred in a query, but the nested column extractors of them are treated differently when we deal with nested column pruning.

### Does this PR introduce _any_ user-facing change?

Yes, fixing a bug in nested column pruning.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @maropu @dongjoon-hyun @frankyin-factual

**Comment by frankyin-factual** [comment]
Great! Thanks. I gonna apply the same patch to my branch.

**Comment by viirya** [comment]
You can rebase after this gets merged.

**Comment by viirya** [comment]
Does Jenkins not work again?

**Comment by viirya** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:157]: Nice catch, @viirya @frankyin-factual !

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:503]: This issue cannot happen in branch-3.0? 
https://github.com/apache/spark/blob/fc2660c302b0c83a9a8a5bec3cc7ae28f8fecdd6/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala#L80-L85

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:182]: Looks we don't need the cast here?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:182]: This part is related to the query failure in the test? Looks it is just an optimization?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:182]: btw, we cannot avoid the generating duplicated aliases (`name#1.first AS _gen_alias_52#52` and `name#1.first AS _gen_alias_53#53`) below? Is this technically difficult? (This is not related to this PR and just a question)
```
scala> sql("select name.first from contact_alias").explain()
== Physical Plan ==
*(2) Project [_gen_alias_52#52 AS first#50]
+- Exchange hashpartitioning(_gen_alias_53#53, _gen_alias_54#54, 100), false, [id=#46]
   +- *(1) Project [name#1.first AS _gen_alias_52#52, name#1.first AS _gen_alias_53#53, name#1.last AS _gen_alias_54#54]
      +- FileScan parquet [name#1,p#8] Batched: false, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/tmp/contacts], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<name:struct<first:string,last:string>>
```

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:182]: Due to cosmetic variations, Extractors with different qualifiers, for example, will cause incorrect total number of fields.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:503]: This test case is only for master branch. However this issue can happen in branch-3.0 too. I added another new test here, which is for branch-3.0.

But when we backport this to branch-3.0, we need to remove first test case it will fail on `checkScan(query, "struct<name:struct<first:string,last:string>>")`, because branch-3.0 doesn't prune for repartition by expression.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:516]: @maropu This test is for branch-3.0.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:182]: For the duplicated aliases, we can avoid it. We can work on canonicalized extractors when generating aliases.

But we also need to convert coming extractors to canonicalized versions when we look up into the alias map.

Currently the code looks clear. And it seems not a big deal, and I think it is rare case that there are multiple extractors with cosmetic difference. So currently I don't try to do that.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:503]: Yea, thanks for adding the test.

**Comment by maropu** [comment]
Looks okay and anyone could check this? @dongjoon-hyun @dbtsai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: > For the duplicated aliases, we can avoid it. We can work on canonicalized extractors when generating aliases.
But we also need to convert coming extractors to canonicalized versions when we look up into the alias map.
Currently the code looks clear. And it seems not a big deal, and I think it is rare case that there are multiple extractors with cosmetic difference. So currently I don't try to do that.

Thanks for the explanation. Looks okay.

**Comment by viirya** [comment]
retest this please

**Comment by maropu** [comment]
FYI: @zhengruifeng @ScrapCodes (Not sure which one is a release manager though) I think this fix's better to be included in the v3.0.1 release.

**Comment by viirya** [comment]
retest this please...

**Comment by viirya** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
The Jenkins looks unstable.

**Comment by maropu** [comment]
Yeah, Shane is working hard on this issue now, so we need to wait a little until it stabilizes.

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala:182]: nit, indentation?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:516]: Thank you for the test case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: fixed.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you so much, @viirya , @maropu , @frankyin-factual , @HyukjinKwon .
Merged to master.

The last commit is a only-indentation fix.

**Comment by dongjoon-hyun** [comment]
Could you adjust the test case and make a backporting PR against branch-3.0, @viirya?

cc @dbtsai

**Comment by viirya** [comment]
Thanks all. Sure, let me create a backporting PR for branch-3.0.

**Pull request closed without merging**
